perf/pkeyread.c

```diff
@@ -0,0 +1,337 @@
+            fprintf(stderr, "Unknown key format (%s)\n", key_format);
+            fprintf(stderr, "key must be:");
+            for (i = 0; format_names[i] != NULL; i++)
+                fprintf(stderr, "\t%s\n", format_names[i]);
+            return EXIT_FAILURE;
+        }
+    }
+
+    if (key_format == NULL) {
+        fprintf(stderr, "option -f is missing\n");
+        return EXIT_FAILURE;
+    }
+
+    if (key == NULL) {
+        fprintf(stderr, "option -k is missing\n");
+        return EXIT_FAILURE;
+    }
+
+    if (threadcount < 1) {
+        fprintf(stderr, "threadcount must be > 0, use option -t 1\n");
+        return EXIT_FAILURE;
+    }
+
+    times = OPENSSL_malloc(sizeof(OSSL_TIME) * threadcount);
+    if (times == NULL) {
+        printf("Failed to create times array\n");
+        return EXIT_FAILURE;
+    }
+
+    if (key_id == SAMPLE_ALL) {
+        if (format_id == FORMAT_ALL) {
+            /* run all samples for all formats */
+            for (k = 0; k < SAMPLE_ALL; k++) {
+                for (f = 0; f < FORMAT_ALL; f++) {
+                    sample_id = k;
+                    if (!perflib_run_multi_thread_test(do_f[f], threadcount, &duration)) {
+                        fprintf(stderr, "Failed to run the test %s in %s format]\n",
+                                sample_names[k], format_names[f]);
+                        return EXIT_FAILURE;
+                    }
+                    report_result(k, f, verbose);
+                }
+            }
+        } else {
+            /* run all samples for one format (format_id) */
+            for (k = 0; k < SAMPLE_ALL; k++) {
+                sample_id = k;
+                if (!perflib_run_multi_thread_test(do_f[format_id], threadcount, &duration)) {
+                    fprintf(stderr, "Failed to run the test %s in %s format]\n",
+                            sample_names[k], format_names[format_id]);
+                    return EXIT_FAILURE;
+                }
+		report_result(k, format_id, verbose);
+            }
+        }
+    } else if (format_id == FORMAT_ALL) {
+        /* run one sample for all formats */
+        for (f = 0; f < FORMAT_ALL; f++) {
+            sample_id = key_id;
+            if (!perflib_run_multi_thread_test(do_f[f], threadcount, &duration)) {
+                fprintf(stderr, "Failed to run the test %s in %s format]\n",
+                        sample_names[key_id], format_names[f]);
+                return EXIT_FAILURE;
+            }
+            report_result(key_id, f, verbose);
+        }
+    } else {
+        /* run one sample in desired format */
+        sample_id = key_id;
+        if (!perflib_run_multi_thread_test(do_f[format_id], threadcount, &duration)) {
+            fprintf(stderr, "Failed to run the test %s in %s format]\n",
+                    sample_names[key_id], format_names[format_id]);
+            return EXIT_FAILURE;
+        }
+        report_result(key_id, format_id, verbose);
+    }
```

Comment:
There is still a lot of code duplication here. Here's one way you could reduce it (not tested):
```suggestion
    if (key_id == SAMPLE_ALL) {
        key_id_min = 0;
        key_id_max = SAMPLE_ALL;
    } else {
        key_id_min = key_id;
        key_id_max = key_id + 1;
    }
    if (format_id == FORMAT_ALL) {
        format_id_min = 0;
        format_id_max = FORMAT_ALL;
    } else {
        format_id_min = format_id;
        format_id_max = format_id + 1;
    }
    /* run samples/formats as appropriate */
    for (k = key_id_min; k < key_id_max; k++) {
        for (f = format_id_min; f < format_id_max; f++) {
            sample_id = k;
            if (!perflib_run_multi_thread_test(do_f[f], threadcount, &duration)) {
                fprintf(stderr, "Failed to run the test %s in %s format]\n",
                        sample_names[k], format_names[f]);
                return EXIT_FAILURE;
            }
            report_result(k, f, verbose);
        }
    }

```

Reply:
It works and it looks good. I like it very much thanks for the suggestion.